src/simplewallet/simplewallet.cpp

```diff
@@ -91,20 +91,9 @@ typedef cryptonote::simple_wallet sw;
 #define MIN_RING_SIZE 7 // Used to inform user about min ring size -- does not track actual protocol
 
 #define LOCK_IDLE_SCOPE() \
-  bool auto_refresh_enabled = m_auto_refresh_enabled.load(std::memory_order_relaxed); \
-  m_auto_refresh_enabled.store(false, std::memory_order_relaxed); \
-  /* stop any background refresh, and take over */ \
   m_wallet->stop(); \
-  m_idle_mutex.lock(); \
-  while (m_auto_refresh_refreshing) \
-    m_idle_cond.notify_one(); \
-  m_idle_mutex.unlock(); \
-/*  if (auto_refresh_run)*/ \
-    /*m_auto_refresh_thread.join();*/ \
   boost::unique_lock<boost::mutex> lock(m_idle_mutex); \
-  epee::misc_utils::auto_scope_leave_caller scope_exit_handler = epee::misc_utils::create_scope_leave_handler([&](){ \
-    m_auto_refresh_enabled.store(auto_refresh_enabled, std::memory_order_relaxed); \
-  })
+  m_idle_cond.notify_one();
```

Comment:
What prevents the background refresh to start between stop and the lock ?

Reply:
Nothing. Nothing bad would happen, though, as far as I can see; just one more autorefresh would squeeze in and unstop the wallet. And no change compared with the status quo in this regard.

But, now that you point this out, I realize that of course it makes more sense to lock **first** and only then call `stop` because that is a state change of the wallet that probably should not happen if one does not currently "own" the wallet.

I will switch the order of those 2 lines.

Currently the only thing within `wallet2` that reacts to such a "stopping" is a running refresh loop that can't run anymore anyway in the current constellation of simplewallet's threads after one has a lock on the mutex (everybody only refreshes under the protection of a lock), so right now that call does nothing if done after locking.

But of course that could change in the future. Maybe in the future there will also be a need for a way for the wallet to signal back "I am really stopped now", because that may take some time after all.

Reply:
Ah, no, wrong, there is a change compared to the status quo: The refresh thread can squeeze in today as well, but it would release the wallet immediately again instead of doing a refresh because of `m_auto_refresh_enabled` set to false.

Reply:
Your change doesn't fix it.

Reply:
Sorry, I don't understand what you mean: Which problem is not fixed by the new order of the lines, with locking the mutex as the very first action? Or, more in general, what problem do you see there in the order "lock, stop, notify"?

Reply:
Your first step is to lock m_idle_mutex. If the background thread is currently refreshing, you'll just lock up until it's done.

Reply:
Hmm, yes, I see, your are correct.

If you think going further here is a waste of time, or if you just are busy, just tell, I am ready to withdraw my PR - it's not important. But if you bear with me, 2 more questions:

     m_idle_mutex.lock(); \
     while (m_auto_refresh_refreshing) \
       m_idle_cond.notify_one(); \
     m_idle_mutex.unlock(); \

In this construct as it is now, how does `notify_one` ever get executed? I really can't see that happening. `m_auto_refresh_refreshing` only goes true in the idle loop, but we here are only able to lock **after**  one idle loop is through and `m_auto_refresh_refreshing` already went back to false.

Second question: Instead of setting `m_auto_refresh_enabled` to false can't we just rely on the fact that the first thing the idle loop does after coming out of `wallet2.refresh` is **wait**, and wait a long time? So we have more than enough time after `stop` to lock the mutex (in the order of the original PR "stop, lock, notify") without any chance that the idle loop squeezes in - because it waits?

Reply:
First: you seem to be correct.
Second: that's gross, and the code might be changed to not wait last. In fact, in my share-rpc branch, it is the case.